src/main/scala/is/hail/expr/ir/functions/Functions.scala

```diff
@@ -0,0 +1,180 @@
+package is.hail.expr.ir.functions
+
+import is.hail.asm4s._
+import is.hail.expr.{AST, FunType, Lambda, TypeTag}
+import is.hail.expr.ir._
+import is.hail.expr.types._
+import is.hail.utils._
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+object IRFunctionRegistry {
+
+  val registry: mutable.Map[String, Seq[(TypeTag, Seq[IR] => IR)]] = mutable.Map().withDefaultValue(Seq.empty)
+
+  UtilFunctions.registerAll(registry)
+
+  def lookupFunction(name: String, args: TypeTag): Option[Seq[IR] => IR] = {
+    val validMethods = registry(name).flatMap { case (tt, f) =>
+      if (tt.xs.length == args.xs.length) {
+        tt.clear()
+        if (tt.unify(args))
+          Some(f)
+        else
+          None
+      }
+      else {
+        None
+      }
+    }
+    validMethods match {
+      case Seq() => None
+      case Seq(x) => Some(x)
+      case _ => fatal(s"Multiple methods found that satisfy $name$args.")
+    }
+  }
+}
+
+abstract class RegistryFunctions {
+
+  def registerAll(r: mutable.Map[String, Seq[(TypeTag, Seq[IR] => IR)]]) {
+    registry.foreach { case (k, v) =>
+      r.put(k, r(k) ++ registry(k))
+    }
+  }
+
+  private val registry = mutable.Map[String, Seq[(TypeTag, Seq[IR] => IR)]]().withDefaultValue(Seq.empty)
```

Comment:
I'm not sure why you need a separate registry.  Why not just add them to IRFunctionRegistry.registry?